### PR TITLE
Implement tool call streaming for AI protocol V5

### PR DESCRIPTION
**Pull request by nvie**
## Summary
- Upgraded AI protocol from V4 to V5 to support tool call streaming
- Implemented real-time streaming of tool invocations as they're being built
- Added comprehensive delta handling for text, reasoning, and tool invocation updates
- Created test page in e2e/next-ai-kitchen-sink for testing tool call streaming functionality

## Key Changes
- **Protocol upgrade**: Bumped AI WebSocket endpoint from `/ai/v4` to `/ai/v5`
- **Delta processing**: Implemented `patchContentWithDelta` function to handle streaming updates
- **Tool streaming**: Added support for `tool-stream`, `tool-delta`, and progressive tool invocation updates
- **JSON parsing**: Added `parsePartialJsonObject` utility for handling incomplete JSON during streaming
- **UI improvements**: Fixed spinner behavior during tool call receiving state
- **Test infrastructure**: Added comprehensive test suite with 30+ test cases covering all streaming scenarios

## Technical Implementation
- Tool invocations now progress through stages: `receiving` → `executing` → `executed`
- Partial JSON arguments are built progressively as deltas arrive
- Content updates are applied in-place to maintain React state consistency
- Backwards compatible with existing non-streaming tool implementations

## Test plan
- [x] All existing tests pass
- [x] New comprehensive test suite covers delta handling edge cases
- [x] Manual testing with tool call streaming test page
- [x] Verified streaming works with complex nested JSON tool arguments

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: packages/liveblocks-core/src/types/ai.ts [hunk 1]
```diff
@@ -484,7 +512,22 @@ export type AiKnowledgeSource = {
 
 // --------------------------------------------------------------------------------------------------
 
-export function appendDelta(
+/**
+ * Polyfill for Array.prototype.findLastIndex()
+ */
+export function findLastIndex<T>(
```
Comment (marcbouchenoire) [inline packages/liveblocks-core/src/types/ai.ts]: We have [another of these in `@liveblocks/react-ui`](https://github.com/liveblocks/liveblocks/blob/main/packages/liveblocks-react-ui/src/utils/find-last-index.ts) but [browser support for it is really good these days](https://caniuse.com/mdn-javascript_builtins_array_findlastindex), should we bump our bundler configs to ES2023?

Comment (nvie) [reply packages/liveblocks-core/src/types/ai.ts]: Ah yes, that would indeed be nice!

File: packages/liveblocks-core/src/lib/parsePartialJsonObject.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    result += '"';
+  }
+
+  // 2. Trim whitespace at the end
+  result = result.trimEnd();
+
+  // 3. If the last char is a ',', strip it
+  if (result.endsWith(",")) {
+    result = result.slice(0, -1);
+  }
```
Comment (nimeshnayaju) [inline packages/liveblocks-core/src/lib/parsePartialJsonObject.ts:47]: Does this only strip the comma if it's the last character? Does that mean `{ "key": "value", }'` (and similar other partial jsons) will be parsed as `{}`?

**Review by nimeshnayaju — APPROVED** [review]

File: packages/liveblocks-core/src/types/ai.ts (hunk 1 above)
Comment (nvie) [reply packages/liveblocks-core/src/types/ai.ts]: I tried "just" enabling es2023, but it became a bit of a rabbit hole of having to upgrade esbuild/tsup/etc to support es2023, so I might do this in a later/dedicated PR instead 😬 For now, I've just DRYed up the polyfills.

File: packages/liveblocks-core/src/lib/parsePartialJsonObject.ts (hunk 2 above)
Comment (nvie) [reply packages/liveblocks-core/src/lib/parsePartialJsonObject.ts:47]: Yes, but that would not be a valid JSON prefix string. This will not happen in practice.

| Input                | Output             |
| -------------------- | ------------------ |
| `{`                  | `{}`               |
| `{"`                 | `{}`               |
| `{"k`                | `{}`               |
| `{"key"`             | `{}`               |
| `{"key": `           | `{}`               |
| `{"key": "`          | `{"key": ""}`      |
| `{"key": "v`         | `{"key": "v"}`     |
| `{"key": "value"`    | `{"key": "value"}` |
| `{"key": "value", `  | `{"key": "value"}` |
| `{"key": "value", }` | `{}` (parse error) |

Comment (nvie) [reply packages/liveblocks-core/src/lib/parsePartialJsonObject.ts:47]: I've added an extensive test suite for this exact behavior, too.
**Update**: Now with [100% test coverage](https://github.com/liveblocks/liveblocks/blob/186379e90787aed6e08e8aaea375be821f9cbdc4/packages/liveblocks-core/src/lib/__tests__/parsePartialJsonObject.test.ts) even.
